Source/WebCore/css/CSSCounterStyle.cpp

```diff
@@ -27,52 +27,211 @@
 #include "CSSCounterStyle.h"
 
 #include "CSSCounterStyleDescriptors.h"
+#include "CSSCounterStyleRegistry.h"
 #include "CSSCounterStyleRule.h"
 #include "CSSPrimitiveValue.h"
 #include "CSSValuePair.h"
+#include <cmath>
+#include <wtf/text/StringBuilder.h>
+#include <wtf/text/TextBreakIterator.h>
 
 namespace WebCore {
 
-String CSSCounterStyle::counterForSystemCyclic(int)
+// https://www.w3.org/TR/css-counter-styles-3/#cyclic-system
+String CSSCounterStyle::counterForSystemCyclic(int value)
 {
-    // FIXME: implement counter for system rdar://103648354
-    return ""_s;
+    auto amountOfSymbols = symbols().size();
+    int symbolIndex = (value - 1) % amountOfSymbols;
+    if (symbolIndex < 0)
+        symbolIndex += amountOfSymbols;
+    if (symbolIndex < 0 || static_cast<unsigned>(symbolIndex) >= amountOfSymbols) {
+        ASSERT_NOT_REACHED();
+        return  { };
+    }
+    return symbols().at(static_cast<unsigned>(symbolIndex));
 }
-String CSSCounterStyle::counterForSystemFixed(int)
+
+// https://www.w3.org/TR/css-counter-styles-3/#fixed-system
+String CSSCounterStyle::counterForSystemFixed(int value)
 {
-    // FIXME: implement counter for system rdar://103648354
-    return ""_s;
+    // Empty string will force value to be handled by fallback.
+    if (value < firstSymbolValueForFixedSystem())
+        return { };
+    unsigned valueOffset = value - firstSymbolValueForFixedSystem();
+    if (valueOffset >= symbols().size())
+        return { };
+    return symbols().at(valueOffset);
 }
-String CSSCounterStyle::counterForSystemSymbolic(int)
+
+// https://www.w3.org/TR/css-counter-styles-3/#symbolic-system
+String CSSCounterStyle::counterForSystemSymbolic(int value)
 {
-    // FIXME: implement counter for system rdar://103648354
-    return ""_s;
+    auto amountOfSymbols = symbols().size();
+    if (!amountOfSymbols || !value) {
```

Comment:
if author uses extends system, it can't define symbols, as defined by the spec. It should inherit the symbols from whatever it is extending. If no symbols are defined, the parser should invalidate the rule. Some systems even require at least 2 symbols, for example.